python/revng/model/_util.py

```diff
@@ -30,3 +30,20 @@ def init_forcing_value(self, *args, **kwargs):
         original_init(self, *args, **kwargs)
 
     base_class.__init__ = init_forcing_value  # type: ignore
+
+
+# NOTE: remove after upgrade to python 3.10 (which introduces kw_ony in dataclasses)
+def force_kw_only(base_class: type):
+    """Monkeypatches the __init__ method so that only kwargs are passed, raises ValueError if args
+    contains values
+    """
+    assert isinstance(base_class, type)
+
+    original_init = base_class.__init__  # type: ignore
+
+    def init_check_kw_only(self, *args, **kwargs):
+        if args is not None and len(args) > 0:
+            raise ValueError(f"Non-keyword arguments passed to constructor: {args}")
+        original_init(self, *args, **kwargs)
+
+    base_class.__init__ = init_check_kw_only  # type: ignore
```

Comment:
Currently Pylance/Pyright/PyCharm do not take into account the modified constructor, so all of them will report the constructor of all types as a list of Positional arguments (instead of properly indicating kw-only arguments), however both of them will correctly report the argument names (e.g. writing `Binary(Fun` will trigger autocomplete for `Functions=`)

However even with python 3.10 neither of them report incorrect initialization when using positional arguments.
The feature can be reworked so that 3.10+ users would see the correct initializer, while both 3.9- and 3.10+ get an error at runtime when using non-kw arguments.

The main problem with leaving positional arguments is that it takes away a lot of the predictability. For instance if a new optional field was added to `Type` it would break all inititalizations of itself and child classes, also since type autogenerates the `Kind` and `ID`, which are the first two fields, it would lead to bad ergonomics IMHO.